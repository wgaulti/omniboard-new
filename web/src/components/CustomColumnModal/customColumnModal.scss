.delete-icon {
  color: #000;
  font-size: 1.2em;
  padding-top: 10px;
  opacity: 0.2;
  &:hover {
    opacity: 0.5;
    cursor: pointer;
  }
}