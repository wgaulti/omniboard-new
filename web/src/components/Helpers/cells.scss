.editable-cell {
  margin-top: -15px;
  margin-left: -8px;
}

.select-cell {
  margin-top: -15px;
  margin-left: -5px;
}

.delete-icon {
  &:hover {
    color: red;
  }
}

.selection-cell-wrapper {
  .checkbox label {
    padding-left: 0;
    line-height: 0;
  }

  .checkbox label:after {
    content: '';
    display: table;
    clear: both;
  }

  .checkbox .cr {
    position: absolute;
    top: 50%;
    left: 50%;
    display: inline-block;
    border: 1px solid #a9a9a9;
    border-radius: .30em;
    width: 1.5em;
    height: 1.5em;
    float: left;
    margin: -0.75em 0 0 -0.75em; /* Half of height 0 0 Half of width */
  }

  .checkbox .cr .cr-icon, .checkbox .cr .cr-indeterminate-icon {
    position: absolute;
    font-size: 0.9em;
    line-height: 0;
    top: 50%;
    left: 15%;
  }

  .checkbox label input[type="checkbox"] {
    display: none;
  }

  .checkbox label input[type="checkbox"] + .cr {
    & > .cr-icon, & > .cr-indeterminate-icon {
      opacity: 0;
    }
  }

  .checkbox label input[type="checkbox"]:checked + .cr {
    background-color: #337ab7;
    color: #FFFFFF;
    border: 1px #FFFFFF;

    & > .cr-icon {
      opacity: 1;
    }
  }

  .checkbox label input[type="checkbox"]:indeterminate + .cr {
    background-color: #337ab7;
    color: #FFFFFF;
    border: 1px #FFFFFF;

    & > .cr-icon {
      opacity: 0;
    }

    & > .cr-indeterminate-icon {
      opacity: 1;
    }
  }

  .checkbox label input[type="checkbox"]:disabled + .cr {
    opacity: .5;
  }
}
