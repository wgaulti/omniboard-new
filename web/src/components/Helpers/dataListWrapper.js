class DataListWrapper {
  constructor(data = [], size, end, fetchCallback) {
    this._data = data;
    // When fetching we need to fetch the index missing + additional x-elements.
    // This specifies that we add 10% of the total size when fetching, the
    // maximum number of data-requests will then be 10.
    this._fetchSize = DataListWrapper._getFetchSize(size);
    this._end = end;
    this._pending = false;
    this._size = size;
    this._fetchCallback = fetchCallback;
  }

  static _getFetchSize(size) {
    return isNaN(size) ? 0 : Math.ceil(Number(size) / 10);
  }

  getSize() {
    return this._size || 0;
  }

  fetchRange(end) {
    if (this._pending) {
      return;
    }

    this._pending = true;
    this._fetchCallback(end)
      .then(({data, count}) => {
        this._pending = false;
        this._end = end;
        this._size = count;
        this._data = data;
        this._fetchSize = DataListWrapper._getFetchSize(count);
      })
      .catch(error => {
        this._pending = false;
        console.error(error);
      });
  }

  getObjectAt(index) {
    // Fetch data 15 rows ahead of end.
    // This will improve user experience such that the user
    // will not see pending status for a long time.
    const numRowsAhead = 15;
    if (index + numRowsAhead >= this._end && this._end < this.getSize()) {
      this.fetchRange(Math.min(this.getSize(),
        index + numRowsAhead + this._fetchSize));
      return index < this._end ? this._data[index] : null;
    }

    return this._data[index];
  }

  getDataArray() {
    return this._data;
  }

  setObjectAt(index, value) {
    this._data[index] = value;
  }

  set data(value) {
    this._data = value;
  }
}

export {DataListWrapper};
