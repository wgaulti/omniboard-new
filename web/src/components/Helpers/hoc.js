import PropTypes from 'prop-types';
import React from 'react';
import './hoc.scss';

export class ProgressWrapper extends React.PureComponent {
  static propTypes = {
    children: PropTypes.node,
    loading: PropTypes.bool
  };

  render() {
    const {loading, children} = this.props;
    let html = children;
    if (loading) {
      html = (
        <div className='table-display'>
          <div className='table-cell-display gray'>
            <i className='glyphicon glyphicon-refresh glyphicon-refresh-animate'/> Loading...
          </div>
        </div>
      );
    }

    return html;
  }
}
