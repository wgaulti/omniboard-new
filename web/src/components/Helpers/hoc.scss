.table-display {
  display: table;
  width: 100%;
  height: 100%;
  .table-cell-display {
    display: table-cell;
    text-align: center;
    vertical-align: middle;
  }
}

.gray {
  color: #808080;
}
