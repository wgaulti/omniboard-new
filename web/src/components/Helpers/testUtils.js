import mockAxios from 'jest-mock-axios';

export const generateMockResponse = (status, times) => {
  for (let i = 0; i < times; i++) {
    mockAxios.mockResponse({status});
  }
};
