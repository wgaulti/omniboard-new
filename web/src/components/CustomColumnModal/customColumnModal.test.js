import React from 'react';
import mockAxios from 'jest-mock-axios';
import {parseServerError} from '../Helpers/utils';
import {CustomColumnModal} from './customColumnModal';

describe('CustomColumnModal', () => {
  let wrapper = null;
  const closeHandler = jest.fn(() => wrapper.setProps({shouldShow: false}));
  const dataUpdateHandler = jest.fn();
  const deleteHandler = jest.fn();

  const responseData = [
    {_id: '5c16204663dfd3fe6a193610', name: 'batch_size', config_path: 'config.train.batch_size', __v: 0},
    {_id: '5c16ea82bea682411d7c0405', name: 'settings_epochs', config_path: 'config.train.settings.epochs', __v: 0},
    {_id: '5c16ebd6bea682411d7c0407', name: 'experiment_name', config_path: 'experiment.name', __v: 0}
  ];

  const runsConfigResponse = [
    {message: 'Hello world!', recipient: 'world', seed: 631323961, train: {batch_size: 32, epochs: 100, lr: 0.01}},
    {seed: 637090657, recipient: 'world', message: 'Hello world!', train: {batch_size: 32, settings: {epochs: 12}, epochs: 100, lr: 0.01}}
  ];

  const hostResponse = [{hostname: 'viveks-imac.lan', os: ['Darwin', 'Darwin-18.6.0-x86_64-i386-64bit'], python_version: '3.7.4', cpu: 'AMD Ryzen Six-Core Processor', ENV: {}}];
  const experimentResponse = [{name: 'hello_config', base_dir: '/Users/vivekratnavel/Documents/sacred_experiment'}];

  beforeEach(() => {
    wrapper = shallow(
      <CustomColumnModal shouldShow handleClose={closeHandler} handleDataUpdate={dataUpdateHandler} handleDelete={deleteHandler}/>
    );
  });

  afterEach(() => {
    // Cleaning up the mess left behind the previous test
    mockAxios.reset();
    jest.clearAllMocks();
  });

  it('should render correctly', async () => {
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.state().isLoadingColumns).toBeTruthy();
    expect(wrapper.state().isLoadingConfigs).toBeTruthy();
    mockAxios.mockResponse({status: 200, data: responseData});
    mockAxios.mockResponse({status: 200, data: runsConfigResponse});
    mockAxios.mockResponse({status: 200, data: hostResponse});
    mockAxios.mockResponse({status: 200, data: experimentResponse});

    await tick();
    expect(wrapper.state().isLoadingColumns).toBeFalsy();
    expect(wrapper.state().isLoadingConfigs).toBeFalsy();
    expect(wrapper.update()).toMatchSnapshot();
  });

  it('should fetch config columns on mount', async () => {
    mockAxios.mockResponse({status: 200, data: responseData});

    expect(mockAxios.get).toHaveBeenCalledWith('api/v1/Runs', {params: {distinct: 'config'}});
    expect(wrapper.state().isLoadingConfigs).toBeTruthy();
    mockAxios.mockResponse({status: 200, data: runsConfigResponse});
    mockAxios.mockResponse({status: 200, data: hostResponse});
    mockAxios.mockResponse({status: 200, data: experimentResponse});

    await tick();
    expect(wrapper.state().isLoadingConfigs).toBeFalsy();
    expect(wrapper.state().configPaths).toHaveLength(16);
    expect(wrapper.state().configPaths).toEqual(
      expect.arrayContaining([
        'config.train',
        'config.train.epochs',
        'config.train.settings.epochs',
        'host.hostname',
        'experiment.base_dir'
      ])
    );
  });

  it('should handle close correctly', () => {
    wrapper.find('[test-attr="close-btn"]').simulate('click');

    expect(closeHandler).toHaveBeenCalledWith();
  });

  it('should handle data update correctly', done => {
    mockAxios.mockResponse({status: 200, data: responseData});
    mockAxios.mockResponse({status: 200, data: runsConfigResponse});
    mockAxios.mockResponse({status: 200, data: hostResponse});
    mockAxios.mockResponse({status: 200, data: experimentResponse});

    expect(wrapper.state().columns[0].id).toEqual(responseData[0]._id);
    expect(wrapper.find('[test-attr="apply-btn"]').props().disabled).toBeTruthy();
    wrapper.find('[test-attr="column-name-text-0"]').simulate('change', {
      target: {
        value: 'col_1'
      }
    });

    expect(wrapper.state().columns[0].columnName).toEqual('col_1');
    // Test if apply button is not disabled
    expect(wrapper.find('[test-attr="apply-btn"]').props().disabled).toBeFalsy();
    wrapper.find('[test-attr="apply-btn"]').simulate('click');

    mockAxios.mockResponse({status: 200, data: [responseData[0]]});

    setTimeout(() => {
      expect(dataUpdateHandler).toHaveBeenCalledWith();
      expect(closeHandler).toHaveBeenCalledWith();
      done();
    });
  });

  describe('should update columns correctly', () => {
    it('and handle create requests', async () => {
      mockAxios.mockResponse({status: 200, data: responseData});
      mockAxios.mockResponse({status: 200, data: runsConfigResponse});
      mockAxios.mockResponse({status: 200, data: hostResponse});
      mockAxios.mockResponse({status: 200, data: experimentResponse});
      wrapper.find('[test-attr="column-name-text-0"]').simulate('change', {
        target: {
          value: 'col_1'
        }
      });
      wrapper.find('[test-attr="add-column-btn"]').simulate('click');
      wrapper.find('[test-attr="column-name-text-3"]').simulate('change', {
        target: {
          value: 'col_4'
        }
      });
      wrapper.find('[test-attr="config-path-3"]').simulate('change', {value: 'train.lr'});

      expect(mockAxios.post).toHaveBeenCalledTimes(0);
      wrapper.find('[test-attr="apply-btn"]').simulate('click');

      expect(mockAxios.post).toHaveBeenCalledTimes(2);
      mockAxios.mockResponse({status: 200, data: []});
      mockAxios.mockResponse({status: 200, data: []});
      await tick();

      expect(wrapper.state().isInProgress).toBeFalsy();
    });
  });

  describe('should handle delete correctly', () => {
    beforeEach(() => {
      mockAxios.mockResponse({status: 200, data: responseData});
      mockAxios.mockResponse({status: 200, data: runsConfigResponse});
      mockAxios.mockResponse({status: 200, data: hostResponse});
      mockAxios.mockResponse({status: 200, data: experimentResponse});
    });
    it('for server side delete', () => {
      wrapper.find('[test-attr="delete-0"]').simulate('click');
      mockAxios.mockResponse({status: 204, data: []});

      expect(deleteHandler).toHaveBeenCalledWith(responseData[0].name);
      expect(wrapper.state().columns).toHaveLength(2);
      expect(wrapper.state().columns[0].id).toEqual(responseData[1]._id);
    });

    it('for client side delete', () => {
      wrapper.find('[test-attr="add-column-btn"]').simulate('click');
      wrapper.find('[test-attr="delete-3"]').simulate('click');

      expect(mockAxios.delete).not.toHaveBeenCalled();
      expect(wrapper.state().columns).toHaveLength(3);
    });
  });

  describe('should handle errors correctly', () => {
    describe('for delete', () => {
      beforeEach(() => {
        mockAxios.mockResponse({status: 200, data: responseData});
        mockAxios.mockResponse({status: 200, data: runsConfigResponse});
        mockAxios.mockResponse({status: 200, data: hostResponse});
        mockAxios.mockResponse({status: 200, data: experimentResponse});
      });
      const errorResponse = {status: 404, message: 'unknown error'};
      it('server error', async () => {
        wrapper.find('[test-attr="delete-0"]').simulate('click');
        mockAxios.mockResponse(errorResponse);
        await tick();

        expect(wrapper.state().columns).toHaveLength(3);
        expect(wrapper.state().error).toEqual(parseServerError(errorResponse));
      });

      it('client error', () => {
        wrapper.find('[test-attr="delete-0"]').simulate('click');
        mockAxios.mockError(errorResponse);

        expect(wrapper.state().error).toEqual(parseServerError(errorResponse));
      });
    });

    it('for fetch config columns error', () => {
      mockAxios.mockError({status: 400, message: 'not found'});

      expect(wrapper.state().error).toEqual(parseServerError({status: 400, message: 'not found'}));
    });

    it('when there are no config paths', async () => {
      mockAxios.mockResponse({status: 200, data: responseData});
      mockAxios.mockResponse({status: 200, data: []});
      mockAxios.mockResponse({status: 200, data: []});
      mockAxios.mockResponse({status: 200, data: []});

      await tick();
      expect(wrapper.state().error).toEqual('There are no nested config parameters available to add a new column');
    });

    describe('for apply', () => {
      beforeEach(() => {
        mockAxios.mockResponse({status: 200, data: responseData});
        mockAxios.mockResponse({status: 200, data: runsConfigResponse});
        mockAxios.mockResponse({status: 200, data: hostResponse});
        mockAxios.mockResponse({status: 200, data: experimentResponse});
      });
      it('client error', () => {
        wrapper.find('[test-attr="apply-btn"]').simulate('click');

        expect(wrapper.state().error).toEqual('There are no changes to be applied');
      });

      it('server error', async () => {
        const errorResponse = {status: 400, response: {data: {message: 'cannot update'}}};
        wrapper.find('[test-attr="config-path-0"]').simulate('change', {value: 'train.epochs'});
        wrapper.find('[test-attr="apply-btn"]').simulate('click');
        mockAxios.mockError(errorResponse);
        await tick();

        expect(wrapper.state().error).toEqual(parseServerError(errorResponse));
      });
    });

    it('for update', async () => {
      const errorResponse = {status: 400, message: 'unknown error'};
      mockAxios.mockResponse({status: 200, data: responseData});
      mockAxios.mockResponse({status: 200, data: runsConfigResponse});
      mockAxios.mockResponse({status: 200, data: hostResponse});
      mockAxios.mockResponse({status: 200, data: experimentResponse});
      wrapper.find('[test-attr="column-name-text-0"]').simulate('change', {
        target: {
          value: 'col_1'
        }
      });
      wrapper.find('[test-attr="apply-btn"]').simulate('click');
      mockAxios.mockResponse(errorResponse);
      await tick();

      expect(wrapper.state().error).toEqual(parseServerError(errorResponse));
    });
  });

  describe('should add columns correctly', () => {
    beforeEach(() => {
      mockAxios.mockResponse({status: 200, data: []});
      mockAxios.mockResponse({status: 200, data: runsConfigResponse});
      mockAxios.mockResponse({status: 200, data: hostResponse});
      mockAxios.mockResponse({status: 200, data: experimentResponse});
      wrapper.find('[test-attr="add-column-btn"]').simulate('click');
      wrapper.find('[test-attr="column-name-text-0"]').simulate('change', {
        target: {
          value: 'col_1'
        }
      });
      wrapper.find('[test-attr="config-path-0"]').simulate('change', {value: 'train.lr'});
      wrapper.find('[test-attr="apply-btn"]').simulate('click');
    });

    afterEach(() => {
      mockAxios.reset();
      jest.clearAllMocks();
    });

    it('and update data handler', async () => {
      mockAxios.mockResponse({status: 201, data: []});
      await tick();

      expect(dataUpdateHandler).toHaveBeenCalledWith();
    });

    it('and handle errors', async () => {
      mockAxios.mockError({status: 400, response: {data: {message: 'cannot create'}}});
      await tick();

      expect(wrapper.state().error).toEqual(parseServerError({status: 400, response: {data: {message: 'cannot create'}}}));
    });
  });

  it('should detect correctly if form is dirty', () => {
    mockAxios.mockResponse({status: 200, data: responseData});
    mockAxios.mockResponse({status: 200, data: runsConfigResponse});
    mockAxios.mockResponse({status: 200, data: hostResponse});
    mockAxios.mockResponse({status: 200, data: experimentResponse});
    wrapper.find('[test-attr="add-column-btn"]').simulate('click');

    expect(wrapper.instance().isFormDirty).toBeTruthy();
  });

  it('should handle config path change correctly', () => {
    mockAxios.mockResponse({status: 200, data: responseData});
    mockAxios.mockResponse({status: 200, data: runsConfigResponse});
    mockAxios.mockResponse({status: 200, data: hostResponse});
    mockAxios.mockResponse({status: 200, data: experimentResponse});
    wrapper.find('[test-attr="config-path-0"]').simulate('change', {value: 'train.epochs'});

    expect(wrapper.state().columns[0].configPath).toEqual('train.epochs');
  });

  it('should reload data when modal dialog is reopened', async () => {
    mockAxios.mockResponse({status: 200, data: responseData});
    mockAxios.mockResponse({status: 200, data: runsConfigResponse});
    mockAxios.mockResponse({status: 200, data: hostResponse});
    mockAxios.mockResponse({status: 200, data: experimentResponse});
    wrapper.find('[test-attr="close-btn"]').simulate('click');

    expect(wrapper.instance().props.shouldShow).toBeFalsy();
    wrapper.setProps({shouldShow: true});

    expect(wrapper.update().state().isLoadingColumns).toBeTruthy();
    expect(wrapper.state().isLoadingConfigs).toBeTruthy();
    expect(mockAxios.get).toHaveBeenCalledWith('api/v1/Runs', {params: {distinct: 'config'}});
  });
});
