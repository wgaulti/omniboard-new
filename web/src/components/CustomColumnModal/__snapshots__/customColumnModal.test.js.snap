// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomColumnModal should render correctly 1`] = `
<Modal
  animation={true}
  autoFocus={true}
  backdrop={true}
  bsClass="modal"
  dialogComponentClass={[Function]}
  enforceFocus={true}
  keyboard={true}
  manager={
    ModalManager {
      "add": [Function],
      "containers": Array [],
      "data": Array [],
      "handleContainerOverflow": true,
      "hideSiblingNodes": true,
      "isTopModal": [Function],
      "modals": Array [],
      "remove": [Function],
    }
  }
  onHide={[MockFunction]}
  renderBackdrop={[Function]}
  restoreFocus={true}
  show={true}
>
  <ModalHeader
    bsClass="modal-header"
    closeButton={true}
    closeLabel="Close"
  >
    <ModalTitle
      bsClass="modal-title"
      componentClass="h4"
    >
      Manage Custom Columns
    </ModalTitle>
  </ModalHeader>
  <ModalBody
    bsClass="modal-body"
    componentClass="div"
  >
    <div
      style={
        Object {
          "paddingBottom": "15px",
        }
      }
    >
      Add Column to expand any nested config / custom parameter into a new column
    </div>
    <ProgressWrapper
      loading={true}
    >
      <div>
        <form />
      </div>
    </ProgressWrapper>
    <div>
      <Button
        active={false}
        block={false}
        bsClass="btn"
        bsSize="small"
        bsStyle="info"
        disabled={false}
        onClick={[Function]}
        test-attr="add-column-btn"
      >
        <span
          aria-hidden="true"
          className="glyphicon glyphicon-plus"
        />
         Add Column
      </Button>
    </div>
  </ModalBody>
  <ModalFooter
    bsClass="modal-footer"
    componentClass="div"
  >
    <Button
      active={false}
      block={false}
      bsClass="btn"
      bsStyle="default"
      disabled={false}
      onClick={[MockFunction]}
      test-attr="close-btn"
    >
      Close
    </Button>
    <Button
      active={false}
      block={false}
      bsClass="btn"
      bsStyle="primary"
      disabled={true}
      onClick={[Function]}
      test-attr="apply-btn"
    >
      <span>
        Apply
      </span>
    </Button>
  </ModalFooter>
</Modal>
`;

exports[`CustomColumnModal should render correctly 2`] = `
<Modal
  animation={true}
  autoFocus={true}
  backdrop={true}
  bsClass="modal"
  dialogComponentClass={[Function]}
  enforceFocus={true}
  keyboard={true}
  manager={
    ModalManager {
      "add": [Function],
      "containers": Array [],
      "data": Array [],
      "handleContainerOverflow": true,
      "hideSiblingNodes": true,
      "isTopModal": [Function],
      "modals": Array [],
      "remove": [Function],
    }
  }
  onHide={[MockFunction]}
  renderBackdrop={[Function]}
  restoreFocus={true}
  show={true}
>
  <ModalHeader
    bsClass="modal-header"
    closeButton={true}
    closeLabel="Close"
  >
    <ModalTitle
      bsClass="modal-title"
      componentClass="h4"
    >
      Manage Custom Columns
    </ModalTitle>
  </ModalHeader>
  <ModalBody
    bsClass="modal-body"
    componentClass="div"
  >
    <div
      style={
        Object {
          "paddingBottom": "15px",
        }
      }
    >
      Add Column to expand any nested config / custom parameter into a new column
    </div>
    <ProgressWrapper
      loading={false}
    >
      <div>
        <form>
          <div
            className="row"
            key="0"
          >
            <div
              className="col col-xs-5"
            >
              <FormGroup
                bsClass="form-group"
                controlId="formColumnName"
                validationState="success"
              >
                <FormControl
                  bsClass="form-control"
                  componentClass="input"
                  onChange={[Function]}
                  placeholder="Enter column name"
                  test-attr="column-name-text-0"
                  type="text"
                  value="batch_size"
                />
              </FormGroup>
            </div>
            <div
              className="col col-xs-5"
            >
              <StateManager
                clearable={false}
                defaultInputValue=""
                defaultMenuIsOpen={false}
                defaultValue={null}
                formatCreateLabel={[Function]}
                isLoading={false}
                onChange={[Function]}
                options={
                  Array [
                    Object {
                      "label": "config.message",
                      "value": "config.message",
                    },
                    Object {
                      "label": "config.recipient",
                      "value": "config.recipient",
                    },
                    Object {
                      "label": "config.seed",
                      "value": "config.seed",
                    },
                    Object {
                      "label": "config.train",
                      "value": "config.train",
                    },
                    Object {
                      "label": "config.train.batch_size",
                      "value": "config.train.batch_size",
                    },
                    Object {
                      "label": "config.train.epochs",
                      "value": "config.train.epochs",
                    },
                    Object {
                      "label": "config.train.lr",
                      "value": "config.train.lr",
                    },
                    Object {
                      "label": "config.train.settings",
                      "value": "config.train.settings",
                    },
                    Object {
                      "label": "config.train.settings.epochs",
                      "value": "config.train.settings.epochs",
                    },
                    Object {
                      "label": "host.hostname",
                      "value": "host.hostname",
                    },
                    Object {
                      "label": "host.os",
                      "value": "host.os",
                    },
                    Object {
                      "label": "host.python_version",
                      "value": "host.python_version",
                    },
                    Object {
                      "label": "host.cpu",
                      "value": "host.cpu",
                    },
                    Object {
                      "label": "host.ENV",
                      "value": "host.ENV",
                    },
                    Object {
                      "label": "experiment.name",
                      "value": "experiment.name",
                    },
                    Object {
                      "label": "experiment.base_dir",
                      "value": "experiment.base_dir",
                    },
                  ]
                }
                placeholder="Path"
                test-attr="config-path-0"
                value={
                  Object {
                    "label": "config.train.batch_size",
                    "value": "config.train.batch_size",
                  }
                }
              />
            </div>
            <div
              className="col col-xs-2"
            >
              <i
                className="glyphicon glyphicon-remove delete-icon"
                onClick={[Function]}
                test-attr="delete-0"
              />
            </div>
          </div>
          <div
            className="row"
            key="1"
          >
            <div
              className="col col-xs-5"
            >
              <FormGroup
                bsClass="form-group"
                controlId="formColumnName"
                validationState="success"
              >
                <FormControl
                  bsClass="form-control"
                  componentClass="input"
                  onChange={[Function]}
                  placeholder="Enter column name"
                  test-attr="column-name-text-1"
                  type="text"
                  value="settings_epochs"
                />
              </FormGroup>
            </div>
            <div
              className="col col-xs-5"
            >
              <StateManager
                clearable={false}
                defaultInputValue=""
                defaultMenuIsOpen={false}
                defaultValue={null}
                formatCreateLabel={[Function]}
                isLoading={false}
                onChange={[Function]}
                options={
                  Array [
                    Object {
                      "label": "config.message",
                      "value": "config.message",
                    },
                    Object {
                      "label": "config.recipient",
                      "value": "config.recipient",
                    },
                    Object {
                      "label": "config.seed",
                      "value": "config.seed",
                    },
                    Object {
                      "label": "config.train",
                      "value": "config.train",
                    },
                    Object {
                      "label": "config.train.batch_size",
                      "value": "config.train.batch_size",
                    },
                    Object {
                      "label": "config.train.epochs",
                      "value": "config.train.epochs",
                    },
                    Object {
                      "label": "config.train.lr",
                      "value": "config.train.lr",
                    },
                    Object {
                      "label": "config.train.settings",
                      "value": "config.train.settings",
                    },
                    Object {
                      "label": "config.train.settings.epochs",
                      "value": "config.train.settings.epochs",
                    },
                    Object {
                      "label": "host.hostname",
                      "value": "host.hostname",
                    },
                    Object {
                      "label": "host.os",
                      "value": "host.os",
                    },
                    Object {
                      "label": "host.python_version",
                      "value": "host.python_version",
                    },
                    Object {
                      "label": "host.cpu",
                      "value": "host.cpu",
                    },
                    Object {
                      "label": "host.ENV",
                      "value": "host.ENV",
                    },
                    Object {
                      "label": "experiment.name",
                      "value": "experiment.name",
                    },
                    Object {
                      "label": "experiment.base_dir",
                      "value": "experiment.base_dir",
                    },
                  ]
                }
                placeholder="Path"
                test-attr="config-path-1"
                value={
                  Object {
                    "label": "config.train.settings.epochs",
                    "value": "config.train.settings.epochs",
                  }
                }
              />
            </div>
            <div
              className="col col-xs-2"
            >
              <i
                className="glyphicon glyphicon-remove delete-icon"
                onClick={[Function]}
                test-attr="delete-1"
              />
            </div>
          </div>
          <div
            className="row"
            key="2"
          >
            <div
              className="col col-xs-5"
            >
              <FormGroup
                bsClass="form-group"
                controlId="formColumnName"
                validationState="success"
              >
                <FormControl
                  bsClass="form-control"
                  componentClass="input"
                  onChange={[Function]}
                  placeholder="Enter column name"
                  test-attr="column-name-text-2"
                  type="text"
                  value="experiment_name"
                />
              </FormGroup>
            </div>
            <div
              className="col col-xs-5"
            >
              <StateManager
                clearable={false}
                defaultInputValue=""
                defaultMenuIsOpen={false}
                defaultValue={null}
                formatCreateLabel={[Function]}
                isLoading={false}
                onChange={[Function]}
                options={
                  Array [
                    Object {
                      "label": "config.message",
                      "value": "config.message",
                    },
                    Object {
                      "label": "config.recipient",
                      "value": "config.recipient",
                    },
                    Object {
                      "label": "config.seed",
                      "value": "config.seed",
                    },
                    Object {
                      "label": "config.train",
                      "value": "config.train",
                    },
                    Object {
                      "label": "config.train.batch_size",
                      "value": "config.train.batch_size",
                    },
                    Object {
                      "label": "config.train.epochs",
                      "value": "config.train.epochs",
                    },
                    Object {
                      "label": "config.train.lr",
                      "value": "config.train.lr",
                    },
                    Object {
                      "label": "config.train.settings",
                      "value": "config.train.settings",
                    },
                    Object {
                      "label": "config.train.settings.epochs",
                      "value": "config.train.settings.epochs",
                    },
                    Object {
                      "label": "host.hostname",
                      "value": "host.hostname",
                    },
                    Object {
                      "label": "host.os",
                      "value": "host.os",
                    },
                    Object {
                      "label": "host.python_version",
                      "value": "host.python_version",
                    },
                    Object {
                      "label": "host.cpu",
                      "value": "host.cpu",
                    },
                    Object {
                      "label": "host.ENV",
                      "value": "host.ENV",
                    },
                    Object {
                      "label": "experiment.name",
                      "value": "experiment.name",
                    },
                    Object {
                      "label": "experiment.base_dir",
                      "value": "experiment.base_dir",
                    },
                  ]
                }
                placeholder="Path"
                test-attr="config-path-2"
                value={
                  Object {
                    "label": "experiment.name",
                    "value": "experiment.name",
                  }
                }
              />
            </div>
            <div
              className="col col-xs-2"
            >
              <i
                className="glyphicon glyphicon-remove delete-icon"
                onClick={[Function]}
                test-attr="delete-2"
              />
            </div>
          </div>
        </form>
      </div>
    </ProgressWrapper>
    <div>
      <Button
        active={false}
        block={false}
        bsClass="btn"
        bsSize="small"
        bsStyle="info"
        disabled={false}
        onClick={[Function]}
        test-attr="add-column-btn"
      >
        <span
          aria-hidden="true"
          className="glyphicon glyphicon-plus"
        />
         Add Column
      </Button>
    </div>
  </ModalBody>
  <ModalFooter
    bsClass="modal-footer"
    componentClass="div"
  >
    <Button
      active={false}
      block={false}
      bsClass="btn"
      bsStyle="default"
      disabled={false}
      onClick={[MockFunction]}
      test-attr="close-btn"
    >
      Close
    </Button>
    <Button
      active={false}
      block={false}
      bsClass="btn"
      bsStyle="primary"
      disabled={true}
      onClick={[Function]}
      test-attr="apply-btn"
    >
      <span>
        Apply
      </span>
    </Button>
  </ModalFooter>
</Modal>
`;
